Bpf policy precedence on deny

jrajahalme opened the pull request:
A lot of the complexity in mapstate generation comes from adding L3/L4 entries to make sure precedence of a L3 deny with a broader/wildcarded L4 proto/port is preserved when there is an L4-only allow rule where the port/proto is a subset of the port/proto of the deny rule: neither rule covers the other due to allow rule having the wildcard identity, but there is an intersection that needs to be denied but would be allowed by the datapath: L3 from the deny rule, the more specific port/proto from the allow rule. We currently manage this by adding a "dependent" L3/L4 deny entry, so that when the datapath selects to most specific of that and the L4-only rule, the added deny rule "wins", as the L4 proto/port is the same but the added rule also has a specific L3.

Alternatively, we could let the datapath manage this intersection by performing both L3 and L4-only policy lookups, if the first match is not a deny. If both lookups find a matching policy, then we simply need to make sure that either one that is a deny rule is selected, regardless of the prefix length of the entries. I.e., only check for the prefix lengths if both entries are allows.

If we apply similar logic to auth type propagation in similar situation, we can remove all the code that relates to adding these dependent L3/L4 keys, as well as the tracking of dependencies between mapstate keys.

The only downside of this is that bpf policy lookup now needs to be done for both L3 and L4-only keys, if the found L3 policy is not a deny. Previously we skipped the 2nd lookup if the match had no port wildcarding (e.g. allow FOO on TCP/80). While I think we now need to do the 2nd lookup more often, especially when deny rules are not used, I think the agent code simplification and some reduction in the policy map size are worth this tradeoff.

This change speeds up BenchmarkRegenerateCIDRDenyPolicyRules by ~10% (~2% more memory used with ~26% less allocations),

```release-note
bpf datapath now manages policy verdict precedence between L3 and wildcard-L3 policy map matches
```

jrajahalme committed:
policy: Use AncestorLongestPrefixFirstIterator for test

Change 'mapStateAllowsKey' to use longest prefix first iterator in
preparation of removal of separate allows and denies maps.

Signed-off-by: Jarno Rajahalme <jarno@isovalent.com>

squeed commented on bpf/lib/policy.h:
```diff
@@ -24,9 +30,23 @@ __account_and_check(struct __ctx_buff *ctx __maybe_unused, struct policy_entry *
-	if (unlikely(policy->auth_type)) {
+
+	auth_type = policy->auth_type;
+	if (unlikely(!policy->has_explicit_auth_type &&
```
Is there a clever way to achieve this with bitwise operations instead of comparisons?

jrajahalme replied:
Not really, `has_explicit_auth_type` is the only bit there, `policy2` is a pointer and `auth_type` is a byte.

jrajahalme committed:
policy: remove mapState.Lookup()

mapState.Lookup is now trivial and only used internally, remove it so we
can use the 'Lookup' name later for policy lookup.

Signed-off-by: Jarno Rajahalme <jarno@isovalent.com>

jrajahalme committed:
policy: Add mapState.String()

Dumping the entries into a string is useful for debugging.

Signed-off-by: Jarno Rajahalme <jarno@isovalent.com>

jrajahalme committed:
policy: Fix auth type merging

Auth type merge logic should select the numerically higher value also
when merging two non-explicit auth types.

Signed-off-by: Jarno Rajahalme <jarno@isovalent.com>

squeed commented on bpf/lib/common.h:
```diff
@@ -398,7 +398,8 @@ struct policy_entry {
 			reserved:2, /* bits used in Cilium 1.16, keep unused for Cilium 1.17 */
 			lpm_prefix_length:5; /* map key protocol and dport prefix length */
-	__u8		auth_type;
+	__u8		auth_type:7,
```
Does this need an update in `cilium-dbg bpf policy list`?

jrajahalme replied:
Yes, fixed!

squeed commented on pkg/policy/mapstate.go:
```diff
@@ -276,6 +276,102 @@ func (ms *mapState) SubsetKeysWithSameID(key Key) iter.Seq2[Key, mapStateEntry]
+// To be used in testing in place of the bpf datapath when full integration testing is not desired.
+// Returns the closest matching covering policy entry and 'true' if found.
+// 'key' must not have a wildcard identity or port.
+func (ms *mapState) Lookup(key Key) (mapStateEntry, bool) {
```
unfortunate that an exported function returns an unexported type. Can this return something simpler? (And maybe an unexported `lookup()` that returns `mapStateEntry`

squeed replied:
though I suppose it's somewhat moot, given that the receiver `mapState` is also unexported.

jrajahalme replied:
Changed this to return the exported MapStateEntry member only. AFAIK this can be used from other packages if there is a function that returns `*mapState` to them, i.e., the receiver type need not be exported and it's exported functions can still be used from other packages.

squeed commented on pkg/policy/mapstate.go:
```diff
@@ -624,19 +624,16 @@ func (e *mapStateEntry) merge(entry *mapStateEntry) {
-				}
-			} else {
-				e.HasAuthType = ExplicitAuthType
+		if entry.HasAuthType == e.HasAuthType {
```
Is this a fix for a previous merged commit, or should this be folded in to the PR?

jrajahalme replied:
I moved this to be the first commit, will post this as a separate PR.

jrajahalme replied:
On second thought, in 1.16 we only ever support one auth type in datapath (SPIRE), so this is not a bug, but future proofing.

squeed commented on pkg/policy/mapstate.go:
```diff
@@ -681,7 +689,7 @@ func (e *mapStateEntry) deepEqual(o *mapStateEntry) bool {
 }
 
-func (e MapStateEntry) WithAuthType(authType AuthType) MapStateEntry {
+func (e MapStateEntry) withAuthType(authType AuthType) MapStateEntry {
```
Does the "unused" lint failure here mean we don't have test cases covering this? Or just that we have created the entries another way?

This is a useful function regardless, feel free to add a `//nolint`

jrajahalme replied:
We now have helpers like `withExplicitAuth()` and `withDerivedAuth()` so this was a real leftover, removed.

jrajahalme committed:
bpf: policy precedence on deny & auth

Let bpf datapath know when an entry has an explicit auth type, as this
can be used to determining whether to override the auth type between
matching policy entries.

When both L4-only and L3/L4 policies match, give precedence to deny
policy. If both are allow policies and that chosen (more specific) policy
has no auth requirement, propagate the auth requirement from the more
generic one.

This change allows removal of creation of dependent L3/L4 policy entries
and hence the dependency tracking between the policy entries as well.

Bpf datapath now takes care of propagating auth type between matching
policy entries so policy package only needs to propagate the excplicit
auth type to covered entries with the same ID.

Signed-off-by: Jarno Rajahalme <jarno@isovalent.com>

jrajahalme committed:
policy: Add policy Lookup() for testing

Add a Lookup function that executes the same precedence logic between
matching L3 and L4-only entries as the bpf datapath does. This allows for
testing "traffic" against the computed endpoint policy without running
the bpf datapath.

Signed-off-by: Jarno Rajahalme <jarno@isovalent.com>

jrajahalme committed:
policy: Add helpers for mapstate entry creation

Add helpers for reducing the boilerplate and call sites for
'newMapStateEntry' itself.

Signed-off-by: Jarno Rajahalme <jarno@isovalent.com>

jrajahalme committed:
policy: Add types/auth.go

Move AuthType and related types to policy/types so that they can be used
without importing the whole policy package.

Add a new AuthRequirement type that defines the high bit as the boolean
for "auth requirement explicitly defined" rather than derived from a
broader key with an explicit auth requirement.

Signed-off-by: Jarno Rajahalme <jarno@isovalent.com>

jrajahalme committed:
policymap: Use policyTypes.AuthRequirement

Use the new policy/types.AuthRequirement type directly in the policymap.

This simplifies wiring the auth requirement from the policy package down
to the bpf datapath policy map.

Signed-off-by: Jarno Rajahalme <jarno@isovalent.com>

jrajahalme committed:
policy: Retire separate allows and denies mapStateMaps

Use a single 'mapStateMap' instead of separate ones for allows and
denies. This is done by deleting the old type 'mapState' and renaming the
old 'mapStateMap' as 'mapState'.

The downsides of this changes are:

- iteration loop bodies become slightly more complicated checking for
  allow/deny
- also allow entries are looped over when looking for a covering deny
  key, which increases memory use by ~15% in
  BenchmarkRegenerateCIDRDenyPolicyRules

These downsides are overcome by ~10% faster operation, partially due to
reduced number of allocations and code simplification due to the removal
of separate maps.

Signed-off-by: Jarno Rajahalme <jarno@isovalent.com>

squeed commented on pkg/policy/mapstate.go:
```diff
@@ -788,19 +788,16 @@ func (e *mapStateEntry) merge(entry *mapStateEntry) {
-				e.hasAuthType = ExplicitAuthType
+		// Numerically higher AuthType takes precedence when both are
+		// either explicitly defined or derived
+		if entry.hasAuthType == e.hasAuthType {
```
can this be replaced with a `max()`? or do we need to track non-set as well?

jrajahalme replied:
We could use `max()` as well for setting the `entry.AuthType` to the higher of the two.

squeed commented on pkg/policy/mapstate_test.go:
```diff
@@ -1571,7 +1535,7 @@ func TestMapState_AccumulateMapChanges(t *testing.T) {
 			egressKey(43, 0, 0, 0):  allowEntry(0, csFoo).WithAuthType(AuthTypeAlwaysFail),
 			egressKey(0, 6, 0, 0):   allowEntry(0, csFoo).WithAuthType(AuthTypeSpire),
-			egressKey(43, 6, 80, 0): allowEntry(1, csBar).WithDefaultAuthType(AuthTypeSpire),
+			egressKey(43, 6, 80, 0): allowEntry(1, csBar).WithDefaultAuthType(AuthTypeAlwaysFail),
```
Is this a bug in the test?

jrajahalme replied:
No, this is due to the auth type now being derived only from the key with the same ID. Now the datapath is responsible for propagating the explicit spire auth type from the L4-only key when both L4-only and L3-keys match.

squeed reviewed (approved):
A few small questions, but looks great!

YutaroHayakawa reviewed (approved):
BPF code looks good to me. I'm not super familiar with the policy code base, but at least, I think this doesn't change the behavior of the existing policy and just add the new criteria (deny and auth) on top of it.
